### PR TITLE
[14.0][IMP] mail_debrand: Keep message body intact while removing branding

### LINKED ISSUE #713: [14.0] mail_debrand: breaks email redirect links in Odoo EE
The regex (`[^(<)(</)]odoo`) in the mail_debrand module seems to be broken.

Firstly - and please correct me if I'm wrong - Python does not support captures like `(xx)` inside groups like `[xx]`. The syntax `[^(<)(</)]` is matching anything that is not `(`, `<`, `/`, or `)`. I believe the intention was to negatively match against `<` or `</`. 

Secondly, this regex _also_ matches any anchor/href containing `odoo`. For example, we have quite a few clients who host Odoo at `https://odoo.company.com`. This regex strips the `/odoo` portion, leaving us with `https:/.company.com`.

Thirdly, the regex is running `.sub()` on the _whole_ match. As above, the match include a single character that is not in `()</`, so it's stripping an extra character preceding `odoo`. Maybe this is the intended behavior?

<hr />

I'd love to make a PR to fix this, but I'm not clear on what the regex is trying to do.

I think maybe `(?<!<)(?<!</)(odoo)(?!\.)` would be a good alternative:
 - Prevents matching if `odoo` is directly preceded by `<` or `</`
 - Prevents matching if `odoo` is directly followed by `.`
 - Does not _match_ the preceding or following part (e.g. `href="odoo" a=1` -> `href="" a=1`, rather than `href= a=1`)

You can test it at on regex101.com [here](https://regex101.com/r/FLTxID/3).

@alexteodor could you clarify a bit on the intent of this regex? Especially, why are we negatively matching `<` and `</`? Also, I'd love to hear your thoughts on the above regex.

### PR DESCRIPTION
**Pull request by joao-p-marques**
Partially fixes https://github.com/OCA/social/issues/713

@Tecnativa
TT31737

ping @pedrobaeza @Yajo @ryanc-me

### CODE REVIEW

File: mail_debrand/tests/test_mail_debrand.py
Comment (pedrobaeza) [inline mail_debrand/tests/test_mail_debrand.py]: Can you please split this by lines (and the other one)?

**Review by pedrobaeza — APPROVED** [review]

File: mail_debrand/tests/test_mail_debrand.py
Comment (joao-p-marques) [reply mail_debrand/tests/test_mail_debrand.py]: Done :+1: Thanks!

File: mail_debrand/models/mail_render_mixin.py
Comment (yajo) [inline mail_debrand/models/mail_render_mixin.py]: I don't understand this comment.

File: mail_debrand/models/mail_render_mixin.py
Comment (yajo) [inline mail_debrand/models/mail_render_mixin.py]: typo?

```suggestion
                odoo_anchors = tree.xpath('//a[contains(@href,"odoo.com")]')
```

File: mail_debrand/models/mail_render_mixin.py
Comment (yajo) [inline mail_debrand/models/mail_render_mixin.py]: As you are processing parts, possibly this is better suited:

```suggestion
                tree = html.fromstring(
```

File: mail_debrand/models/mail_render_mixin.py
Comment (yajo) [inline mail_debrand/models/mail_render_mixin.py]:
```suggestion
                        previous.tail = ""
```

File: mail_debrand/models/mail_render_mixin.py
Comment (yajo) [inline mail_debrand/models/mail_render_mixin.py]: I find it weird that you remove the parent even if you got `remove_parent=False`. Does that mean that we can just remove that argument from the method?

```suggestion
                        if parent.tag == "td":  # also here can be powered by
```

File: mail_debrand/models/mail_render_mixin.py
Comment (yajo) [inline mail_debrand/models/mail_render_mixin.py]: Are there really any `<odoo> whatever </odoo>` elements added to emails? 🤔

File: mail_debrand/models/mail_render_mixin.py
Comment (yajo) [inline mail_debrand/models/mail_render_mixin.py]:
```suggestion
                part = etree.tostring(tree, pretty_print=True, method="html", encoding="unicode")
```

**Review by yajo — CHANGES_REQUESTED** [review]

File: mail_debrand/models/mail_render_mixin.py
Comment (joao-p-marques) [reply mail_debrand/models/mail_render_mixin.py]: Well, I think this regex is actually badly constructed, as @ryanc-me also suggested and tried to fix in https://github.com/OCA/social/pull/714

Right now, it is actually matching (according to https://regexr.com):
- A negated set of:
    - a `(` character
    - a `<` character
    - a `)` character
    - all the above again :sweat_smile: 
    - a `/` character
- The `odoo` word

So, it obviously needs changing... However, as @ryanc-me seems to have a fix ready in the PR mentioned above, I just wanted to fix another part of the problem, and then maybe we can proceed from what he already did...

**Comment by joao-p-marques** [comment]
@Yajo thanks for the review!

Most of your comments actually refer to the previous code that was not my own, so I can't explain it completely :sweat_smile: ... That's why the goal with this PR was not to change it, but just to add this new checks to avoid extra changes... However, I attended most of them, as they where mostly typos and obvious fixes... 

As for the only one left unattended (about the `<odoo>` tags), as explained, maybe it is better to leave it to another PR, especially as there is one already open.

File: mail_debrand/models/mail_render_mixin.py
Comment (ryanc-me) [reply mail_debrand/models/mail_render_mixin.py]: @joao-p-marques you're exactly right about the regex, there's a bit more info in #713. The PR #714 does "fix" that issue (see this [regex101](https://regex101.com/r/o01Rfz/5)), but I don't think the regex approach is actually correct. It leaves us with `<span>Powered by <a></a></span>`, which is *technically* de-branding, but isn't a nice solution.

I'm not convinced we can remove all branding with a generic regex - maybe instead we need to find all strings like "Odoo" in the  CE/EE sources, and write some specific rules?


Also @Yajo, just for reference, I think the original author was trying not to match the `<odoo></odoo>` tags surrounding all Odoo templates (the top-level element in all XML docs)

Comment (joao-p-marques) [reply mail_debrand/models/mail_render_mixin.py]: Thanks for the explanation!

Well, with this PR, as we are making sure to leave the message body intact, we can be a bit more "strict" while reformatting the rest, which already solves a big part of the problem... From my testing, even with this wrong regex, we are being able to actually hide what we want without removing too much, so that takes some urgency out of the other part.

Once this is merged, maybe you can rebase your PR so that we can keep the discussion there and try out your changes.

Comment (yajo) [reply mail_debrand/models/mail_render_mixin.py]: > From my testing, even with this wrong regex, we are being able to actually hide what we want

Yes, that's what the code tells me. AFAICS we're already detecting the "odoo" parts properly and removing the garbage. This regex is useless now (it could be before too, but now it certainly is), so please just remove it.

In case there's some corner case not being detected now, then a future PR can fix it and provide a valid test for it.

Comment (joao-p-marques) [reply mail_debrand/models/mail_render_mixin.py]: Ok, removed then. Let's work on reviewing and #714 to make sure we don't miss anything and don't introduce any other errors...

**Comment by joao-p-marques** [comment]
Changes done @Yajo

**Review by yajo — APPROVED** [review]

**Comment by joao-p-marques** [comment]
The bot is broken :sweat_smile: https://twitter.com/PedroMBaeza/status/1438405500236963841?s=20

**Pull request merged**
